NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +45,206 @@ function Base.show(io::IO, s::CategoryProduct)
   return print(io, ")")
 end
 
-category_show(io::IO, k, v) = print(io, v)
-
+category_show(io::IO, ::Int, v) = print(io, v)
 category_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return categories_isless(categories(s1), categories(s2))
+end
+function Base.isless(s1::CategoryProduct, s2::CategoryProduct)
+  return categories_isless(categories(s1), categories(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for CategoryProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+function categories_fusion_rule(cats1, cats2)
+  diff_cat = CategoryProduct(find_diff(cats1, cats2))
+  shared1, shared2 = find_common(cats1, cats2)
+  fused = map(fusion_rule, values(shared1), values(shared2))
+  return recover_key(typeof(shared1), fused) × diff_cat
+end
+
+# get clean results when mixing implementations
+categories_isequal(::Tuple, ::NamedTuple) = false
+categories_isequal(::NamedTuple, ::Tuple) = false
+
+categories_isless(::NamedTuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(::Tuple, ::NamedTuple) = throw(ArgumentError("Not implemented"))
+
+categories_type(::Type{<:CategoryProduct{T}}) where {T} = T
+
+recover_key(T::Type, t::Tuple{Vararg{AbstractCategory}}) = sector(T, t)
+recover_key(T::Type, c::AbstractCategory) = recover_key(T, (c,))
+recover_key(T::Type, c::CategoryProduct) = recover_key(T, categories(c))
+function recover_key(T::Type, fused::Tuple)
+  # here fused contains at leat one GradedUnitRange
+  g0 = reduce(×, fused)
+  # convention: keep unsorted blocklabels as produced by F order loops in ×
+  new_labels = recover_key.(T, GradedAxes.blocklabels(g0))
+  new_blocklengths =
+    LabelledNumbers.labelled.(GradedAxes.unlabel.(BlockArrays.blocklengths(g0)), new_labels)
+  return GradedAxes.gradedrange(new_blocklengths)
+end
```

Comment:
I'm having trouble understanding what `recover_key` means and what purpose it serves, could you explain its purpose? It may need a better name, a more focused definition, etc.

Reply:
@emstoudenmire curious to hear your thoughts on this design.

Reply:
This is the hardest part of `Sectors`. As I mentioned above, fusing two `CategoryProduct` is made by fusing independently their matching categories. This is done inside `categories_fusion_rule` with
```julia
fused = map(fusion_rule, values(shared1), values(shared2))
```
 If the `CategoryProduct` was a `NamedTuple`, one then needs to recover the keys of that tuple, which are stored in the type. However, similarly to `flatten_maybe_nested`, non-abelian groups carry another difficulty:

- If all categories are all abelian, then `fused` is a tuple of `AbstractCategory`, one just needs to recover the correct type (and possibly the `NamedTuple` keys). This is done with
```julia
recover_key(T::Type, t::Tuple{Vararg{AbstractCategory}}) = sector(T, t)
```
- If there is (at least) one non abelian group involved, then  `fused` actually contains (at least) one `GradedOneTo`.This will call
```julia
recover_key(T::Type, fused::Tuple)
```
This method  loops over all combinations of sectors within the unit range(s) and take the Cartesian product: this is done by `g0 = reduce(×, fused)`.

Then, one needs to recover the exact `CategoryProduct` type on each of these new `CategoryProduct`: done by `new_labels = recover_key.(T, GradedAxes.blocklabels(g0))`. `blocklabels` may contain a `CategoryProduct` and call `recover_key(T::Type, c::CategoryProduct)` or just one `AbstractCategory` and call `recover_key(T::Type, c::AbstractCategory)`.

Finally, one recovers the _structural_  multiplicities from `g0` and outputs a `GradedOneTo` that has the correct multiplicities and the correct labels.

Reply:
This does sound quite complicated. I think we should focus on simplifying this code logic, I strongly suspect this can be simplified quite a bit.

If `fusion_rule` outputs different formats depending on whether the category/group is abelian or non-abelian, what about defining a wrapper function, say `gradedrange_fusion_rule`, that calls `fusion_rule` but always converts the output to a graded unit range? I.e., my suggestion would be to try to "canonicalize" the data formats as much as possible across those different situations so this higher level generic function can be simpler.

I'm also still confused about the name `recover_key`, I see that somehow it has to do with dealing generically with Tuple and NamedTuple formats but maybe we can just rethink how that is handled, maybe that just requires different implementations for those two cases since making one generic function that handles both of those seems like it is getting pretty complicated and hard to understand. I'm not certain that is the way to go, but I think the current way is just too hard to understand and maybe splitting it apart between those two cases at first can then help with coming up with a better generic/combined approach later on.

One thing to consider is that we also want to be able to support data formats like `Vector` and `Dict` inside `CategoryProduct`, where there is less information at the type level. Thinking about that limit may be useful, if a function is only made generic between `Tuple` and `NamedTuple` but not those other formats, I'm not sure if it is worth making generic (or considering those other formats may inspire a different, simpler way of making a combined function that is potentially generic across those formats, say assuming they implement some minimal number of functions).

Reply:
> If `fusion_rule` outputs different formats depending on whether the category/group is abelian or non-abelian, what about defining a wrapper function, say `gradedrange_fusion_rule`, that calls `fusion_rule` but always converts the output to a graded unit range? I.e., my suggestion would be to try to "canonicalize" the data formats as much as possible across those different situations so this higher level generic function can be simpler. **EDIT:** That may apply to canonicalizing the output of `fused = map(fusion_rule, values(shared1), values(shared2))` as well, i.e. construct a conversion function `fused = canonicalize_fused(map(fusion_rule, values(shared1), values(shared2))` that ensures `fused` is always the same data structure independent of the inputs.

We precisely want `fusion_rule` output type to depend on the input. We want `sector(U1(1)×U1(0)) ⊗ sector(U1(1)×U1(0))` to be an `AbstractCategory` as well as `sector(U1(1)×SU2(0)) ⊗ sector(U1(1)×SU2(0))` to be a `GradedOneTo{CategoryProduct{U1,SU2}}`.

Abelian groups aside, assuming `fused` elements are all cast to a `GradedOneTo`, the difficulty lingers. It is required to convert e.g. `Tuple{GradedOneTo{SU2},GradedOneTo{SU2}}` to  `GradedOneTo{CategoryProduct{Tuple{SU2,SU2}}`, or maybe `GradedOneTo{CategoryProduct{NamedTuple{A=SU2,B=SU2}}`. This is one way of doing it, I guess I should use different function names for "recover CategoryProduct from categories" and "convert Tuple of GradedAxes to GradedAxes of Tuple".

Here the lowest level of `recover_key` was tailored for `NamedTuple`. It happens to work for `Tuple` too, therefore I kept generic code. I am open to more generic `CategoryProduct`, but that was never mentioned previously, hence I was not trying to support it.

Reply:
> We precisely want `fusion_rule` output type to depend on the input. We want `sector(U1(1)×U1(0)) ⊗ sector(U1(1)×U1(0))` to be an `AbstractCategory` as well as `sector(U1(1)×SU2(0)) ⊗ sector(U1(1)×SU2(0))` to be a `GradedOneTo{CategoryProduct{U1,SU2}}`.

You misunderstood what I wrote, please try to read more carefully. I was suggesting making a new function that wraps `fusion_rule` but makes sure the output is always `GradedOneTo`, I was not suggesting that you should change the behavior of `fusion_rule`.

> Abelian groups aside, assuming `fused` elements are all cast to a `GradedOneTo`, the difficulty lingers. It is required to convert e.g. `Tuple{GradedOneTo{SU2},GradedOneTo{SU2}}` to `GradedOneTo{CategoryProduct{Tuple{SU2,SU2}}`, or maybe `GradedOneTo{CategoryProduct{NamedTuple{A=SU2,B=SU2}}`. This is one way of doing it, I guess I should use different function names for "recover CategoryProduct from categories" and "convert Tuple of GradedAxes to GradedAxes of Tuple".

I think using better function names would be a good start, so that we can get on the same page about this functionality and how to improve it. Please do your best to rewrite it in a way that is simpler to understand, and we can discuss it again after that given the feedback I've given you. This code can't be merged as it is, it has to be written in a way that we can understand and maintain it.

> Here the lowest level of `recover_key` was tailored for `NamedTuple`. It happens to work for `Tuple` too, therefore I kept generic code. I am open to more generic `CategoryProduct`, but that was never mentioned previously, hence I was not trying to support it.

I'm not saying that you need to support those other formats explicitly, but thinking about more general cases could guide you towards writing better generic code that is more understandable (I find that can be helpful when I am designing generic code). If you can rewrite it in a way that is simpler and easier to understand, I'm fine if it handles both `Tuple` and `NamedTuple`, but your message made it sound like supporting both was one reason why this function is complicated.

Reply:
Please have a look at the latest version of `categories_fusion_rule`.

Reply:
Thanks for the rewrite, I left some new comments based on the latest code, so I'll close this thread.